backup: use ioctx.ReadAll when reading index metadata

kev-cao opened the pull request:
We currently use `io.Reader`'s `Read` to read the index metadata file. However, `Read` behavior w.r.t to returning an EOF after reading the entire file is not consistent --- S3's implementation returns EOF even if bytes are read, whereas GCS's implementation returns EOF only if a read is attempted after reading the file. The correct solution is to use `ioctx.ReadAll` whenever performing reads.

Fixes: #153215

Fixes: #153214 

Fixes: #153213 

Fixes: #153212

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/153369)
<!-- Reviewable:end -->

kev-cao committed:
backup: use ioctx.ReadAll when reading index metadata

We currently use `io.Reader`'s `Read` to read the index metadata file.
However, `Read` behavior w.r.t to returning an EOF after reading the
entire file is not consistent --- S3's implementation returns EOF even
if bytes are read, whereas GCS's implementation returns EOF only if a
read is attempted after reading the file. The correct solution is to use
`ioctx.ReadAll` whenever performing reads.

Fixes: #153215, #153214, #153213, #153212

Release note: None

kev-cao commented:
TFTR!

bors r=msbutler
